tests/bdd/test_volume_create.py

```diff
@@ -137,6 +159,44 @@ def there_are_no_available_mayastor_instances():
     container.kill()
 
 
+@then("there should not be any specs relating to the volume")
+def there_should_not_be_any_specs_relating_to_the_volume():
+    """there should not be any specs relating to the volume."""
+    # Restart the core agent so that all the specs are reloaded from the persistent store.
+    global specs
+    docker_client = docker.from_env()
+    core = docker_client.containers.get("core")
+    core.restart()
+
+    # After restarting the core container it can take a little time to refresh the specs,
+    # so retry a number of times.
+    for i in range(5):
+        try:
+            specs = common.get_specs_api().get_specs()
+            break
+        except:
+            time.sleep(1)
+
+    assert len(specs["volumes"]) == 0
+    assert len(specs["nexuses"]) == 0
+    assert len(specs["replicas"]) == 0
+
+
+@then("volume creation should fail")
+def volume_creation_should_fail(create_request):
+    """volume creation should fail."""
+    request = create_request[CREATE_REQUEST_KEY]
+    try:
+        common.get_volumes_api().put_volume(VOLUME_UUID, request)
+    except Exception as e:
+        exception_info = e.__dict__
+        assert exception_info["status"] == requests.codes["precondition_failed"]
```

Comment:
Is 'precondition failed' here a strong indication of the timeout ?
Can we 100% be sure that we have experienced the operation timeout rather than a general failure due to some pre-conditions, like: we have no available mayastor nodes so the request completes immediately, hence no timeout in general.
Shall we measure time somehow in the test to detect the duration of the operation or so ?

Reply:
I think for the time being this is the best we can do. The alternative is to check the exact error string but this isn't ideal because if the error message changes slightly all the tests would fall apart.

In this case this is a strong indication of the timeout because the previous step ensures that the mayastor instances are deleted and therefore cannot be reached.